src/pcre2_context.c

```diff
@@ -409,6 +410,29 @@ ccontext->stack_guard_data = user_data;
 return 0;
 }
 
+PCRE2_EXP_DEFN int PCRE2_CALL_CONVENTION
+pcre2_set_optimize(pcre2_compile_context *ccontext, uint32_t directive)
+{
+if (directive == PCRE2_OPTIMIZATION_NONE)
+  {
+  ccontext->optimization_flags = 0;
+  }
+else if (directive == PCRE2_OPTIMIZATION_FULL)
+  {
+  ccontext->optimization_flags = PCRE2_OPTIMIZATION_ALL;
+  }
+else if (directive >= PCRE2_AUTO_POSSESS && directive <= PCRE2_START_OPTIMIZE_OFF)
+  {
+  /* Even directive numbers switch a bit on, odd numbers switch a bit off.
+   * 64-65 affect the LSB, 66-67 the 2 bit, 68-69 the 4 bit, and so on. */
+  if (directive & 0x1)
+    ccontext->optimization_flags &= ~(1u << ((directive >> 1) - 32));
+  else
+    ccontext->optimization_flags |= 1u << ((directive >> 1) - 32);
+  }
+
+return 0;
```

Comment:
We should return an error code, when the optimization is not known.

Reply:
I thought about this and deliberately made `pcre2_set_optimize` ignore unknown directives... the reason was so that programs can potentially still work if compiled with headers from a newer version of PCRE2, but linked with an older version.

Of course, any program which uses `pcre2_set_optimize` needs to link with at least PCRE2 10.45 or higher. But after that, let's imagine that 10.53 adds some amazing new optimization called "fleeble squozzilization". Some users don't want this, so they write `pcre2_set_optimize(ccontext, PCRE2_FLEEBLE_SQUOZZILIZE_OFF)`. If such programs are linked with PCRE2 10.45-10.52, the directive will be unknown. It seemed to me that there is no harm in older versions of PCRE2 simply ignoring the unknown directive, since 1) it doesn't implement the unknown optimization anyways, and 2) in any case, turning optimizations on and off shouldn't change observable behavior.

Thinking about it again, the flaw in my reasoning is that users *can* still ignore the error code even if we return one. But it might be useful information to know that the current version of PCRE2 doesn't understand a certain directive.

The other flaw is that some existing optimizations *do* change observable behavior, and that is one reason why we allow people to turn them off. Though I think going forward, any more optimizations like this should be opt-in, not opt-out.

Reply:
OK, this has been changed.